### PR TITLE
[f5_bigip_cookie_disclosure] Store cookies in the database

### PR DESCRIPTION
**Pull request by SkypLabs**
Hi,

This PR follows the feature request https://github.com/rapid7/metasploit-framework/issues/12187.

Additionally, the references have been updated and style issues have been fixed based on the output of [rubocop](https://github.com/rubocop-hq/ruby-style-guide).

## Verification

- [x] Start `msfconsole` and make sure to use a database
- [x] `use auxiliary/gather/f5_bigip_cookie_disclosure`
- [x] `set RHOSTS <vulnerable IP>`
- [x] Verify that the module returns something like `[+] F5 BigIP load balancing cookie...`
- [x] `notes` and verify that the cookie name is stored in the notes

### CODE REVIEW

**Comment by SkypLabs** [comment]
@space-r7 The documentation has been added :+1:

File: documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md
Comment (h00die) [inline documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md]: I think this line can be removed, everyone always does something similar to this.

File: documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md
Comment (h00die) [inline documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md]: One of the big adds to this module was storing the cookie.  Can you do `notes` or however to show that they were stored?

**Review by h00die — CHANGES_REQUESTED** [review]

File: documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md
Comment (SkypLabs) [reply documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md]: Sure thing. Done.

File: documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md
Comment (SkypLabs) [reply documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md]: Good point. Will do.

Comment (SkypLabs) [reply documentation/modules/auxiliary/gather/f5_bigip_cookie_disclosure.md]: Done :+1:

**Comment by h00die** [comment]
I'm happy with the docs, i'll leave it to someone @ r7 to see if they have access to a F5 to test the updates against.

**Review by h00die — APPROVED** [review]
docs look good to me

**Comment by SkypLabs** [comment]
Hello,

Any news on this PR?

If you struggle getting your hands on an F5 BigIP, you can use Shodan or similar to find some online. The module will only collect and parse web cookies. Nothing harmful / illegal.

**Comment by ccondon-r7** [comment]
Hey @h00die, this still in the works? Let us know if you want us to take over instead!

**Comment by h00die** [comment]
Yea I don't have that device or legal authority to use/scan one. You guys may have more luck

**Comment by ccondon-r7** [comment]
Thanks, @h00die!

**Comment by ccondon-r7** [comment]
@SkypLabs we can't test this, unfortunately, so unless someone in the community has (legal) access to a device to test the module, we won't be able to merge it. We added our `needs testing` label in hopes of enlisting community help.

**Comment by smcintyre-r7** [comment]
Since we can't merge this without testing it and we don't have access to (and legal authorization) for testing this vulnerability, I'm got to mark this as attic and close it out. If at a later point in time we can get a PCap, or access and authorization for testing we'd be happy to revisit it at that time.

Thanks alot for the contribution!

**Comment by SkypLabs** [comment]
Hi @ccondon-r7, @smcintyre-r7,

This is unfortunate but I understand the situation. I will try to get some evidence of the good functioning of my patch using a [virtual appliance](https://www.f5.com/trials/big-ip-virtual-edition).

I will come back to you once done. Until then, stay safe!

**Comment by h00die** [comment]
@SkypLabs Must have missed your comment on the virtual appliance.  I have one in my lab, but no idea how to configure it to be vulnerable.  If you can provide some instructions on this, I can get it tested and landed!

**Comment by SkypLabs** [comment]
@h00die Great! Let me check the documentation and I will come back to you with instructions. Many thanks for your help!

**Comment by SkypLabs** [comment]
It seems that all you need is described on [this page](https://techdocs.f5.com/en-us/bigip-15-0-0/big-ip-local-traffic-manager-implementations/configuring-http-load-balancing-with-cookie-persistence.html) in the section "Creating a custom cookie persistence profile".

**Comment by h00die** [comment]
@SkypLabs I found my VM, its a completely default stock image.  I set the cookie persistence profile exactly as the instructions said.  Then I scanned the F5 web interface, but got "F5 BIG-IP load balancing cookie not found".  I'm pretty sure I'm not supposed to scan the F5 web interface directly, however I don't know enough about the devices to know what i need to set up networking wise to make the cookie happen.
image

Any insight for me?

**Comment by h00die** [comment]
@SkypLabs any help?

**Comment by SkypLabs** [comment]
Hey @h00die, sorry for the delay.

Indeed, you won't get the persistent cookies by scanning the admin interface. You need to create a target pool of web servers to process the HTTP incoming requests with your custom cookie persistence profile attached to it. Then, you need to create a virtual server which has the previous created pool as target. You will get the cookies by scanning the IP of the virtual server.

You will find step-by-step instructions in the same [online documentation](https://techdocs.f5.com/en-us/bigip-15-0-0/big-ip-local-traffic-manager-implementations/configuring-http-load-balancing-with-cookie-persistence.html) as the one you followed to create the cookie persistence profile.

Once again, many thanks for your precious help!

**Comment by h00die** [comment]
going through some backlog, i was able to verify this... yes, 1.5yrs later.  I'll do some updates to the docs and such to get it back up to date for 2022

**Comment by h00die** [comment]
https://github.com/SkypLabs/metasploit-framework/pull/2

**Comment by SkypLabs** [comment]
Hi @h00die.

Many thanks for your help! I just merged your PR.

Let me know if you need me for anything.

**Comment by h00die** [comment]
https://www.shodan.io/search?query=%22Set-Cookie%3A+BIGipServer%22

```
msf6 auxiliary(gather/f5_bigip_cookie_disclosure) > run
[*] Running module against 111.111.111.111

[*] Starting request /
[+] F5 BIG-IP load balancing cookie "BIGipServer~thing~thing2~thing3 = rd1o00000000000000000000ffffac292158o73" found
[+] Load balancing pool name "~thing~thing2~thing3" found
[+] Route domain "1" found
[+] Backend 172.11.11.11:80 found
[*] Auxiliary module execution completed
msf6 auxiliary(gather/f5_bigip_cookie_disclosure) > notes

Notes
=====

 Time                     Host             Service  Port  Protocol  Type                           Data
 ----                     ----             -------  ----  --------  ----                           ----
 2022-01-08 17:20:21 UTC  111.111.111.111                           f5_load_balancer_cookie_name   "BIGipServer~thing~thing2~thing3"
 2022-01-08 17:20:21 UTC  111.111.111.111                           f5_load_balancer_pool_name     "~thing~thing2~thing3"
 2022-01-08 17:20:21 UTC  111.111.111.111                           f5_load_balancer_route_domain  "1"
 2022-01-08 17:20:21 UTC  111.111.111.111                           f5_load_balancer_backends      [{:host=>"172.11.11.11", :port=>80}]
```


working with the updates!

**Comment by h00die** [comment]
Merged in https://github.com/rapid7/metasploit-framework/commit/7b11429b49c7e8cdc703d6a4f7ccd376a3237c8d

**Comment by h00die** [comment]
# Release Notes
This PR adds the f5 load balancer cookie to notes, and cleans up the module (rubocop/documentation/refs)

**Comment by SkypLabs** [comment]
Great news. Thanks again @h00die.

**Pull request closed without merging**
